src/dmd/ctfe/bc.d

```diff
@@ -0,0 +1,2369 @@
+*/
+    }
+
+    BCFunction endFunction()
+    {
+        //assert(insideFunction);
+        //I have no idea how this can fail ...
+        localCount = 0;
+
+        insideFunction = false;
+        BCFunction result;
+        result.type = BCFunctionTypeEnum.Bytecode;
+        result.maxStackUsed = sp;
+        result.fn = functionId;
+        {
+            // MUTEX BEGIN
+            // result.byteCode = byteCodeArray[4 .. ip];
+            // MUTEX END
+        }
+        sp = StackAddr(4);
+
+        // import std.stdio; writeln("bc: ", printInstructions(byteCodeArray[4 .. ip]));
+
+        return result;
+    }
+
+    BCValue genParameter(BCType bct, string name = null)
+    {
+        auto p = BCValue(BCParameter(++parameterCount, bct, sp));
+        p.name = name;
+        // everything that's not a basic type is a pointer
+        // and therefore has size 4
+        auto pSize = 4;
+
+        if (isBasicBCType(bct))
+        {
+            pSize = align4(basicTypeSize(bct.type));
+        }
+
+        assert(pSize == 4 || pSize == 8, "we only support 4byte and 8byte params: " ~ enumToString(bct.type));
+
+        sp += pSize;
+
+        return p;
+    }
+
+    BCAddr beginJmp()
+    {
+        BCAddr atIp = ip;
+        ip += 2;
+        return atIp;
+    }
+
+    StackAddr currSp()
+    {
+        return sp;
+    }
+
+    void LoadFramePointer(BCValue intoHere, int offset)
+    {
+        emitLongInst(LongInst.LoadFramePointer, intoHere.stackAddr, imm32(offset).imm32);
+    }
+
+    void endJmp(BCAddr atIp, BCLabel target)
+    {
+        auto offset = isShortJump(target.addr - atIp);
+        if (offset)
+        {
+            byteCodeArray[atIp] = ShortInst16(LongInst.RelJmp, offset);
+        }
+        else
+        {
+            byteCodeArray[atIp] = LongInst.Jmp;
+            byteCodeArray[atIp + 1] = target.addr;
+        }
+    }
+
+    BCLabel genLabel()
+    {
+        return BCLabel(ip);
+    }
+
+    CndJmpBegin beginCndJmp(BCValue cond = BCValue.init, bool ifTrue = false)
+    {
+        if (cond.vType == BCValueType.Immediate)
+        {
+            cond = pushOntoStack(cond);
+        }
+
+        auto result = CndJmpBegin(ip, cond, ifTrue);
+        ip += 2;
+        return result;
+    }
+
+    void endCndJmp(CndJmpBegin jmp, BCLabel target)
+    {
+        auto atIp = jmp.at;
+        auto cond = jmp.cond;
+        auto ifTrue = jmp.ifTrue;
+
+        LongInst64 lj;
+
+        if (isStackValueOrParameter(cond))
+        {
+            lj = (ifTrue ? LongInst64(LongInst.JmpNZ, cond.stackAddr,
+                target.addr) : LongInst64(LongInst.JmpZ, cond.stackAddr, target.addr));
+        }
+        else // if (cond == bcLastCond)
+        {
+            lj = (ifTrue ? LongInst64(LongInst.JmpTrue,
+                target.addr) : LongInst64(LongInst.JmpFalse, target.addr));
+        }
+
+        byteCodeArray[atIp] = lj.lw;
+        byteCodeArray[atIp + 1] = lj.hi;
+    }
+
+    void Jmp(BCLabel target)
```

Comment:
(style nit) methods should begin with lower case and this should probably be named `emitJmp`